py4cast/io/outputs.py

```diff
@@ -86,70 +87,160 @@ def save_named_tensors_to_grib(
-                    target_ds[name] = (dims, data2grib)
-                    target_ds[name] = target_ds[name].assign_attrs(
-                        **model_ds[group][name].attrs
-                    )
             filename = get_output_filename(saving_settings, sample, leadtimes[t_idx])
             option = (
                 "wb"
                 if not os.path.exists(f"{saving_settings.directory}/{filename}")
                 else "ab"
             )
             xtg.to_grib(
-                target_ds,
+                receiver_ds,
                 Path(saving_settings.directory) / filename,
                 option,
             )
 
 
+def write_template_dataset(
+    pred: NamedTensor,
+    ds: DatasetABC,
+    template_ds: xr.Dataset,
+    group: str,
+    sample: Any,
+    validtime: float,
+    leadtime: float,
+    raw_data: NamedTensor,
+    grib_features: pd.DataFrame,
+) -> xr.Dataset:
+    """Write the template xarray dataset with raw data tensor from inference
+
+    Args:
+        pred (NamedTensor): complete namedtensor, containing feature names
+        ds (DatasetABC): inference dataset
+        template_ds (xr.Dataset): xarray dataset extracted from the template grib
+        group (str): index of the template_ds in the template dict containing coherent groups
+        sample (Any): the inference sample to be saved
+        validtime (float): time of validity of the current sample
+        leadtime (float): lead time of the current sample
+        raw_data (NamedTensor): extraction from pred at current timestep
+        grib_features (pd.DataFrame): complete description of feature names and definition
+
+    Returns:
+        xr.Dataset: the template dataset, filled with data from raw_data, in the correct format
+    """
+    receiver_ds = deepcopy(template_ds)
+
+    # if the shape of the dataset grid doesn't match grib template, fill the rest of the data with NaNs
+    nanmask, latlon = make_nan_mask(ds, receiver_ds)
+    (
+        latmin,
+        latmax,
+        longmin,
+        longmax,
+    ) = latlon
+
+    receiver_ds["time"] = sample.date
+    ns_step = np.timedelta64(
+        int(leadtime * 3600 * 1000000000),
+        "ns",
+    )
+    ns_valid = np.timedelta64(
+        int(validtime * 3600 * 1000000000),
+        "ns",
+    )
+    receiver_ds["step"] = ns_step
+    receiver_ds["valid_time"] = np.datetime64(sample.date) + ns_valid
+
+    # retrieving key metadata to be able to parallelize writing
+    namelist = list(receiver_ds.keys())
+    used_grib_feat = grib_features[(grib_features["name"].isin(namelist))]
+
+    # will only be used if namelist has a single item
+    name = namelist[0]
+    feature_names = used_grib_feat["feature_name"].tolist()
+    tol = used_grib_feat["typeOfLevel"].drop_duplicates().tolist()[0]
+    feature_idx = torch.tensor([pred.feature_names_to_idx[f] for f in feature_names])
+
+    data = (
+        raw_data.index_select_dim("features", feature_idx, bare_tensor=True)
```

Comment:
I think bare_tensor=True is the default value, you could get rid of it.

Reply:
done